### DISCUSSION TITLE
How to access Nuxt components from Nitro

### DISCUSSION THREAD

**Discussion by Dorianslavov** [nuxt/nuxt#27502] [Questions] [answered]
We have a full-stack application, Nuxt on the frontend and Nitro on the backend ( the backend is in nuxt's `/server` directory)

We have a PDF service that creates a headless browser, populates it with HTML and returns a downloadable PDF.
For this PDF's template we have a `Preview.vue` component which is used on the frontend and intended to be used also on the backend paired with Vue's exposed functionality which compiles `.vue` files to plain html like that 
```ts
import { renderToString } from 'vue/server-renderer'
import { createSSRApp } from 'vue'

/** Compiles Vue component into pure HTML string */
const renderToHTML = async (component: Component, props: Record<string, any>) => {
    const ssrApp = createSSRApp(component, props)
    const html = await renderToString(ssrApp)

    return html
}
```

The problem with all of this is that inside of Nuxt's `/server` directory, we don't have access to Nuxt imports such as `#components` so I can't just import the `Preview.vue`component and run it through the code snippet from above.
Even if I manage to import it somehow I will also probably run into problems such as that Nitro can't understand `.vue` files which will require me to add manually a Vue plugin to `nitro.rollupConfig` inside of `nuxt.config.ts` which in a way reinvents the wheel.

What would be the best way to solve this problem ? In the end all I want is to have a template `.vue` file which works on the frontend and also can be compiled to plain html so the backend PDF service can use that HTML and populate the PDF with it.

**Comment by RizkiNfs** [comment]
I have a similar problem. I want to generate HTML using `renderToString` from .vue components. After digging into the Nuxt codebase, I don't think it's possible to import .vue files inside the `/server` directory without adding the VuePlugin to `nitro.rollupConfig`. My guess the .vue files are processed inside the Nuxt part of the app before Nitro. Adding VuePlugin allows me to import .vue files, but I don't know how to make auto-import works.

**Reply by Dorianslavov** [reply]
What did you actually do to manage to import `.vue` files inside `/server` directory ? Which Vue plugin did you add to `rollupConfig` and did you set it up in some specific way ? Also how do you import the files, could you copy-paste an example, please ?

**Reply by RizkiNfs** [reply, marked as answer]
nuxt.config.ts
```
import PluginVue from '@vitejs/plugin-vue'
export default defineNuxtConfig({
  nitro: {
    rollupConfig: {
      plugins: [
        PluginVue()
      ]
    }
  },
})
```

server/route or server/api handler

```
import { createSSRApp, h } from 'vue'
import { renderToString } from 'vue/server-renderer'
import UiButton from '~/components/ui/button/index.vue'

export default defineEventHandler(async (event) => {

  setResponseHeader(event, 'Content-Type', 'text/html')

  const app = createSSRApp({
    render(){
      return h(UiButton, 'click me')
    }
  })

  const html = await renderToString(app)

  return (`
    <html lang="en">
      <body>
        <div>${html}</div>
      </body>
    </html>
  `)
})
```

**Reply by Dorianslavov** [reply]
Awesome thanks!
I seem to be hitting a problem with `.vue` files which have postcss styles inside of them like this

![image](https://github.com/nuxt/nuxt/assets/25179716/455131ad-8a8f-4fcc-b4a8-076e93bc02e8)

I have added a rollup postcss plugin to `rollupConfig` but no success
```ts
import vuePlugin from 'rollup-plugin-vue'
import postcssPlugin from 'rollup-plugin-postcss'

...


rollupConfig: {
        plugins: [
                vuePlugin({ include: /\.vue$/, target: 'node' }),
                postcssPlugin(),
        ]
}
```
            
Have you managed to solve the PostCSS issue or you are not experiencing it at all ?

**Reply by Dorianslavov** [reply]
Also FYI, you can work around the auto-import issue which you are experiencing by manually importing the components/utils/composables

**Reply by RizkiNfs** [reply]
I tried to replicate your problem, even with vanilla css I got the same error
```vue
<template>
  <button class="button">
    <slot />
  </button>
</template>

<style>
  .button {
    padding: 1rem;
  }
</style>
```
` RollupError: components/ui/button/index.vue?vue&type=style&index=0&lang.css (2:0): Expression expected (Note that you need plugins to import files that are not JavaScript)`


It's hard for me to explain, but the error makes sense. Even if there is no error, `renderToString` only return the HTML, how would you extract CSS from a .vue file simply by importing it? I think there is a missing step



In my project I'm not using `<style>` block for styling, but using tailwindcss 
this how I generate css in my project

```js
import { createSSRApp, h } from 'vue'
import { renderToString } from 'vue/server-renderer'
import postcss from 'postcss'
import tailwindcss from 'tailwindcss'
import UiButton from '~/components/ui/button/index.vue'

async function generateTailwindCss(html) {
  const { css } = await postcss([
    tailwindcss({
      content: [{ raw:html }],
      theme: {
        extend: {},
      },
      plugins: [],
      corePlugins: {
        preflight: false,
      },
    }), 
  ])
    .process('@tailwind utilities;', { from:undefined })

  return css
}

export default defineEventHandler(async (event) => {

  setResponseHeader(event, 'Content-Type', 'text/html')

  const app = createSSRApp({
    render(){
      return h(UiButton, {class: 'bg-blue-500 px-4 py-2'}, 'click me')
    }
  })

  const html = await renderToString(app)
  const css = await generateTailwindCss(html)

  return (`
    <html lang="en">
      <style>${css}</style>
      <body>
        <div>${html}</div>
      </body>
    </html>
  `)
})
```

**Reply by RizkiNfs** [reply]
I found a workaround,  you can create preview page, and use $fetch to get the HTML
```js
export default defineEventHandler(async (event) => {

  const html = await $fetch('/preview/pdf')

  doSomething(html)  
 
})
```

just make sure all the necessary CSS included in the `<head>`

**Reply by flogehring** [reply]
I'm trying to do something similar for generating a report email through sendgrid within a Nitro scheduled task.
I don't want to put it behind a api-endpoint so it's not exposed to the internet. I have not found a way to mark an event handler as internal only. I think this also wouldn't work for @TheAlexLichter 's suggestion. 
What else could I try?

**Comment by TheAlexLichter** [comment]
I'd argue that you could use Nuxt's Server Components to render the component on the server with the necessary props (exposed through an API route).

**Reply by robokozo** [reply]
I'd be interesting in seeing this kind of approach. I'd like to have an api route dynamically render an SVG image. I'd love to use components to define the svg
